client/src/main/java/io/pravega/client/connection/impl/RawClient.java

```diff
@@ -135,6 +135,12 @@ private void closeConnection(Throwable exceptionToInflightRequests) {
             log.warn("Closing connection to segment {} with exception", segmentId, exceptionToInflightRequests);
         }
         if (closed.compareAndSet(false, true)) {
+            if (connection == null) {
```

Comment:
`private final CompletableFuture<ClientConnection> connection;`
 I do not see how `connection` can be set to null.
@Tristan1900 could you elaborate here please ?

Reply:
sure. It's because the [closeConnection](https://github.com/pravega/pravega/blob/055494a2e2dabfb649e5b7e3c9c4f586dd18940e/client/src/main/java/io/pravega/client/connection/impl/RawClient.java#L130) method is called during the initialization of [connection](https://github.com/pravega/pravega/blob/master/client/src/main/java/io/pravega/client/connection/impl/RawClient.java#L101). The creation of connection requires to register the `Flow` to a `FlowHandler` at [here](https://github.com/pravega/pravega/blob/055494a2e2dabfb649e5b7e3c9c4f586dd18940e/client/src/main/java/io/pravega/client/connection/impl/FlowHandler.java#L88) and that's where the race condition happens. The `Flow` has been registered but the method is not finished so `ClientConnection` does not return and so `RawClient#Connection` is still `Null`. However, at this point, `ClientSocketReader` throws an exception and closes the TcpConnection and the `FlowHandler` is forced to shut down. All registered `Flow` on that FlowHandler is required to shut down and the corresponding callbacks are called, so the `RawClient#closeConnection` is called even when `RawClient#Connection` is not created and that's why `NPE` is thrown.
Hope that makes sense

Reply:
I am not sure we don't have a threading issue here. 
The fact that closeConnection can be called before connection is set means we are working with a partially initialized object, and the fact that a future is involved means we aren't guaranteed this is one thread. 
I think the correct solution is to first assign the future for connection and pass the future into the connect method for it to complete rather than having it return the value. Then there is are no assignments happening, and we don't have to deal with the null case here.

Reply:
Thanks @tkaitchuck, that's a good way to do. I have made the change according to that, could you please take a look again to see if it is what you mean? Thanks!